File: sdks/typescript/src/features/readability.ts
import nlp from 'compromise';
import { syllable } from 'syllable';

/**
 * Flesch-Kincaid grade level over compromise + syllable.
 *
 * Approximates Python's `textstat.flesch_kincaid_grade()` but does not match it, and is not
 * the implementation the contract declares. One caller remains — vocabulary-complexity,
 * which is the sole entry left in PREPROCESSING_GAPS in registry-conformance.test.ts.
 */
export function calculateFleschKincaidGrade(text: string): number {
  return calculateReadabilityMetrics(text).fleschKincaidGrade;
}

/**
 * Additional readability metrics
 */
export interface ReadabilityMetrics {
  sentenceCount: number;
  wordCount: number;
  characterCount: number;
  syllableCount: number;
  avgWordsPerSentence: number;
  avgSyllablesPerWord: number;
  fleschKincaidGrade: number;
}

export function calculateReadabilityMetrics(text: string): ReadabilityMetrics {
  const doc = nlp(text);

  const sentences = doc.sentences().length;
  const terms = doc.terms();
  const words = terms.length;
  const characters = text.replace(/\s/g, '').length;

  const allWords = terms.out('array');
  const totalSyllables = allWords.reduce((sum: number, word: string) => sum + syllable(word), 0);

  const avgWordsPerSentence = sentences > 0 ? words / sentences : 0;
  const avgSyllablesPerWord = words > 0 ? totalSyllables / words : 0;
  const fkGrade = 0.39 * avgWordsPerSentence + 11.8 * avgSyllablesPerWord - 15.59;

  return {
    sentenceCount: sentences,
    wordCount: words,
    characterCount: characters,
    syllableCount: totalSyllables,
    avgWordsPerSentence,
    avgSyllablesPerWord,
    fleschKincaidGrade: Math.round(Math.max(0, fkGrade) * 100) / 100,
  };
}

/**
 * The deterministic counts the sentence-structure contract binds to `{ground_truth_counts}`.
 *
 * The prompt instructs the model to treat these as a reference rather than a constraint, so
 * the block's field names and order are part of the prompt's contract, not a display choice.
 */
export function formatGroundTruthCounts(text: string): string {
  const metrics = calculateReadabilityMetrics(text);

  return [
    `num_sentences: ${metrics.sentenceCount}`,
    `num_words: ${metrics.wordCount}`,
    `num_char: ${metrics.characterCount}`,
    `num_syllable: ${metrics.syllableCount}`,
    `flesch_kincaid_grade: ${metrics.fleschKincaidGrade}`,
  ].join('\n');
}

File: sdks/python/src/learning_commons_evaluators/features/readability.py
"""Deterministic text counts a contract binds to a prompt.

``textstat`` is the library every contract names under ``implementation.python``, so the
counts here are textstat's. They are deliberately not the TypeScript SDK's, which the same
contracts bind to ``compromise`` + ``syllable`` under ``implementation.typescript``: the
registry declares one implementation per language and the two disagree on syllables, so a
cross-SDK prompt comparison masks every value computed here (SDK spec §10.4).
"""

from __future__ import annotations

from dataclasses import dataclass

import textstat

from learning_commons_evaluators.features.preprocessing import format_number, round_half_up


@dataclass(frozen=True)
class ReadabilityCounts:
    """What ``textstat`` counts in one passage."""

    sentence_count: int
    word_count: int
    character_count: int
    syllable_count: int
    #: Rounded to two places by the same rule the contracts' ``post_transform`` applies
    #: to ``fk_score``, so one SDK never rounds one quantity two ways.
    flesch_kincaid_grade: float


def readability_counts(text: str) -> ReadabilityCounts:
    """Count ``text`` the way the sentence-structure notebook counted it.

    Each call is the textstat function the notebook used, with its arguments: words exclude
    punctuation and characters exclude spaces, which is what makes the numbers comparable
    to the ones the prompt's examples were written against.
    """
    return ReadabilityCounts(
        sentence_count=textstat.sentence_count(text),
        word_count=textstat.lexicon_count(text, removepunct=True),
        character_count=textstat.char_count(text, ignore_spaces=True),
        syllable_count=textstat.syllable_count(text),
        flesch_kincaid_grade=round_half_up(textstat.flesch_kincaid_grade(text), 2),
    )


def compute_ground_truth_counts(text: str) -> str:
    """The block the sentence-structure contract binds to ``{ground_truth_counts}``.

    The prompt tells the model to treat these as a reference rather than a constraint, and
    reads them as ``name: value`` lines, so the field names and their order are part of the
    prompt's contract rather than a display choice.
    """
    counts = readability_counts(text)
    return "\n".join(
        [
            f"num_sentences: {counts.sentence_count}",
            f"num_words: {counts.word_count}",
            f"num_char: {counts.character_count}",
            f"num_syllable: {counts.syllable_count}",
            f"flesch_kincaid_grade: {format_number(counts.flesch_kincaid_grade)}",
        ]
    )


__all__ = ["ReadabilityCounts", "compute_ground_truth_counts", "readability_counts"]
